src/security-server/admin-service/application/src/main/java/org/niis/xroad/securityserver/restapi/service/AcmeService.java

```diff
@@ -0,0 +1,341 @@
+        } else if (base64StringWithoutPadding.length() % 4 == 0) {
+            return base64StringWithoutPadding;
+        } else {
+            throw new ValidationFailureException(EAB_SECRET_LENGTH);
+        }
+    }
+
+    private Order createOrder(String commonName, String subjectAltName, Account account) throws AcmeException {
+        log.debug("Creating new order");
+        return account.newOrder()
+                .domains(subjectAltName != null ? subjectAltName : commonName)
+                .notAfter(Instant.now().plus(Duration.ofDays(1L)))
+                .create();
+    }
+
+    private void doAuthorizationAndFinalizeOrder(byte[] certRequest, Order order)
+            throws AcmeException {
+        log.debug("Starting authorization");
+        for (Authorization auth : order.getAuthorizations()) {
+            if (auth.getStatus() == Status.PENDING) {
+                Http01Challenge httpChallenge = auth.findChallenge(Http01Challenge.class)
+                        .orElseThrow(() -> new AcmeServiceException(HTTP_CHALLENGE_MISSING));
+                String token = httpChallenge.getToken();
+                String content = httpChallenge.getAuthorization();
+                String acmeChallenge = acmeChallengePath + token;
+                try {
+                    AtomicSave.execute(acmeChallenge, "tmp_challenge",
+                            out -> out.write(content.getBytes(StandardCharsets.UTF_8)));
+                } catch (Exception e) {
+                    throw new AcmeServiceException(HTTP_CHALLENGE_FILE_CREATION, e);
+                }
+                try {
+                    httpChallenge.trigger();
+                } catch (AcmeException e) {
+                    throw new AcmeServiceException(CHALLENGE_TRIGGER_FAILURE, e);
+                }
+                waitForTheChallengeToBeCompleted(httpChallenge);
+                try {
+                    Files.delete(Path.of(acmeChallenge));
+                } catch (IOException e) {
+                    throw new AcmeServiceException(HTTP_CHALLENGE_FILE_DELETION, e);
+                }
+                log.debug("Finalizing order");
+                order.execute(certRequest);
+            }
+        }
+    }
+
+    private static void waitForTheChallengeToBeCompleted(Challenge challenge) {
+        log.debug("Waiting for challenge to be completed");
+        int attempts = SystemProperties.getAcmeAuthorizationWaitAttempts();
+        long interval = SystemProperties.getAcmeAuthorizationWaitInterval();
+        while (challenge.getStatus() != Status.VALID  && attempts-- > 0) {
+            if (challenge.getStatus() == Status.INVALID) {
+                throw new AcmeServiceException(AUTHORIZATION_FAILURE);
+            }
+            Instant now = Instant.now();
+            try {
+                Instant retryAfter = challenge.fetch().orElse(now.plusSeconds(interval));
+                Thread.sleep(now.until(retryAfter, ChronoUnit.MILLIS));
```

Comment:
I know, there are probably never multiple acme challenges running concurrently, but still could this refactored to async code without `Thread.sleep()`?